app/geant-exporter/geant-exporter.cc

```diff
@@ -80,14 +87,176 @@ void store_particles(TFile* root_file, G4ParticleTable* particle_table)
 
         cout << "  Added " << g4_particle_def->GetParticleName() << endl;
     }
+    cout << endl;
+
+    root_file->Write();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write physics table data to ROOT.
+ *
+ * The ROOT file must be open before this call.
+ */
+void store_physics_tables(TFile* root_file, G4ParticleTable* particle_table)
+{
+    REQUIRE(root_file);
+    REQUIRE(particle_table);
+
+    // Start table writer
+    GeantPhysicsTableWriter table_writer(root_file);
+
+    cout << "Exporting physics tables..." << endl;
+
+    G4ParticleTable::G4PTblDicIterator* particle_iterator
+        = G4ParticleTable::GetParticleTable()->GetIterator();
+    particle_iterator->reset();
+
+    while ((*particle_iterator)())
+    {
+        G4ParticleDefinition* g4_particle_def = particle_iterator->value();
+
+        // Skip "dummy" particles: generic ion and geantino
+        if (g4_particle_def->GetPDGEncoding() == 0)
+            continue;
+
+        std::cout << "=============" << std::endl;
+        std::cout << g4_particle_def->GetParticleName() << std::endl;
+        std::cout << "=============" << std::endl;
+
+        G4ProcessVector* process_list
+            = (g4_particle_def->GetProcessManager())->GetProcessList();
+
+        for (std::size_t j = 0; j < process_list->size(); j++)
+        {
+            G4VProcess* process = (*process_list)[j];
+            table_writer.add_physics_tables(process, g4_particle_def);
+        }
+    }
+    cout << endl;
 
     root_file->Write();
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * This application exports particle and physics table data constructed by the
- * selected physics list. Output data is stored into a ROOT file.
+ * Recurring loop over all logical volumes.
+ *
+ * Function called by store_geometry(...)
+ */
+void loop_volumes(GeantGeometryMap& geometry, G4LogicalVolume* logical_volume)
+{
+    GeantVolume              volume;
+    GeantGeometryMap::vol_id volume_id;
+    GeantGeometryMap::mat_id material_id;
+
+    volume.name = logical_volume->GetName();
+    volume_id   = logical_volume->GetInstanceID();
+    material_id = logical_volume->GetMaterialCutsCouple()->GetIndex();
+
+    // Add volume to the global volume map
+    geometry.add_volume(volume_id, volume);
+
+    // Map volume to its material
+    geometry.link_volume_material(volume_id, material_id);
+
+    // Check for daughter volumes
+    if (logical_volume->GetNoDaughters() > 0)
+    {
+        // Repeat for every daughter volume
+        for (size_type i = 0; i < logical_volume->GetNoDaughters(); i++)
+        {
+            auto daughter_vol
+                = logical_volume->GetDaughter(i)->GetLogicalVolume();
+
+            loop_volumes(geometry, daughter_vol);
```

Comment:
I'd add a comment noting that this is recursive. Is it possible for the same logical volume to be referenced by multiple parents? Is it possible to just loop over logical volume IDs instead of recursing like this?

Reply:
I'll dig deeper, but I was not able to fetch a simple full list of volume ids so far.